File: scripts/access_review.py
from __future__ import annotations

import argparse
import datetime as dt
import hashlib
import hmac
import json
import os
import pathlib
import secrets
import socket
import subprocess
import sys
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Any, Callable

ROOT = pathlib.Path(__file__).resolve().parents[1]
SOURCE_FIELDS = (
    "review_id", "source_id", "target_state", "access_basis", "terms_snapshot_at", "evidence_id",
    "access_basis_valid", "privacy_review_pass", "parser_contract_pass", "provenance_hash_pass",
    "rate_budget_pass", "shadow_pass",
)
PROVIDER_FIELDS = (
    "review_id", "provider_id", "access_basis", "terms_snapshot_at", "rate_policy", "evidence_id",
    "access_basis_valid", "terms_review_pass", "rate_policy_review_pass",
)
PROVIDER_ACCESS_BASES = {"OFFICIAL_API", "PARTNER_CONTRACT", "AIRLINE_DIRECT", "MANUAL_ORACLE", "PUBLIC_PAGE_MONITOR"}
ALLOWED_MODES = {"SHADOW_ACCEPTANCE", "PRODUCTION"}


class AccessReviewError(RuntimeError):
    pass


class TransportUnknown(AccessReviewError):
    pass


def _compact(value: Any) -> str:
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))


def _sha256_text(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def _git_head() -> str:
    return subprocess.check_output(["git", "rev-parse", "HEAD"], cwd=ROOT, text=True).strip().lower()


def _require_clean() -> None:
    if subprocess.check_output(["git", "status", "--porcelain"], cwd=ROOT, text=True).strip():
        raise AccessReviewError("ACCESS_REVIEW_DIRTY_WORKTREE")


def _iso(value: Any) -> bool:
    if not isinstance(value, str) or value == "RECHECK_REQUIRED":
        return False
    try:
        return dt.datetime.fromisoformat(value.replace("Z", "+00:00")).tzinfo is not None
    except ValueError:
        return False


def _required_bool(review: dict[str, Any], name: str) -> None:
    if type(review.get(name)) is not bool:
        raise AccessReviewError(f"ACCESS_REVIEW_FIELD_INVALID:{name}")


def validate_record(record: Any) -> dict[str, Any]:
    if not isinstance(record, dict) or set(record) != {"kind", "review"}:
        raise AccessReviewError("ACCESS_REVIEW_WRAPPER_FIELDS_INVALID")
    kind = record.get("kind")
    review = record.get("review")
    if kind not in {"source", "provider"} or not isinstance(review, dict):
        raise AccessReviewError("ACCESS_REVIEW_KIND_INVALID")
    expected = SOURCE_FIELDS if kind == "source" else PROVIDER_FIELDS
    if set(review) != set(expected):
        raise AccessReviewError("ACCESS_REVIEW_FIELDS_INVALID")
    for name in ("review_id", "evidence_id", "terms_snapshot_at"):
        if not isinstance(review.get(name), str) or not review[name]:
            raise AccessReviewError(f"ACCESS_REVIEW_FIELD_INVALID:{name}")
    entity_key = "source_id" if kind == "source" else "provider_id"
    if not isinstance(review.get(entity_key), str) or not review[entity_key]:
        raise AccessReviewError(f"ACCESS_REVIEW_FIELD_INVALID:{entity_key}")
    if not _iso(review["terms_snapshot_at"]):
        raise AccessReviewError("ACCESS_REVIEW_TERMS_TIMESTAMP_INVALID")
    if kind == "source":
        if review["target_state"] not in {"SHADOW", "ENABLED"}:
            raise AccessReviewError("ACCESS_REVIEW_TARGET_STATE_INVALID")
        if not isinstance(review["access_basis"], str) or not review["access_basis"].strip():
            raise AccessReviewError("ACCESS_REVIEW_ACCESS_BASIS_INVALID")
        for name in SOURCE_FIELDS[-6:]:
            _required_bool(review, name)
    else:
        if review["access_basis"] not in PROVIDER_ACCESS_BASES:
            raise AccessReviewError("ACCESS_REVIEW_ACCESS_BASIS_INVALID")
        if not isinstance(review["rate_policy"], str) or not review["rate_policy"].strip():
            raise AccessReviewError("ACCESS_REVIEW_RATE_POLICY_INVALID")
        for name in PROVIDER_FIELDS[-3:]:
            _required_bool(review, name)
    return {"kind": kind, "review": {name: review[name] for name in expected}}


def load_jsonl(path: pathlib.Path) -> list[dict[str, Any]]:
    records: list[dict[str, Any]] = []
    seen_reviews: set[str] = set()
    seen_evidence: set[str] = set()
    for line_no, raw in enumerate(path.read_text(encoding="utf-8").splitlines(), 1):
        if not raw.strip():
            continue
        try:
            record = validate_record(json.loads(raw))
        except json.JSONDecodeError as exc:
            raise AccessReviewError(f"ACCESS_REVIEW_JSON_INVALID:{line_no}") from exc
        review = record["review"]
        if review["review_id"] in seen_reviews:
            raise AccessReviewError(f"ACCESS_REVIEW_DUPLICATE_REVIEW_ID:{review['review_id']}")
        if review["evidence_id"] in seen_evidence:
            raise AccessReviewError(f"ACCESS_REVIEW_DUPLICATE_EVIDENCE_ID:{review['evidence_id']}")
        seen_reviews.add(review["review_id"])
        seen_evidence.add(review["evidence_id"])
        records.append(record)
    if not records:
        raise AccessReviewError("ACCESS_REVIEW_INPUT_EMPTY")
    return records


def review_payload_hash(record: dict[str, Any]) -> str:
    return _sha256_text(_compact(record["review"]))


def _provider_origin_from(data: dict[str, Any]) -> str | None:
    candidates = [data.get("worker_origin"), data.get("base_url"), data.get("origin"), data.get("workers_dev_origin")]
    nested = data.get("provider_readback")
    if isinstance(nested, dict):
        candidates += [nested.get("worker_origin"), nested.get("base_url"), nested.get("origin"), nested.get("workers_dev_origin")]
    for value in candidates:
        if isinstance(value, str) and value.startswith("https://"):
            return value.rstrip("/")
    return None


def resolve_base_url(head: str, explicit: str | None = None, provider_path: pathlib.Path | None = None) -> str:
    provider_path = provider_path or ROOT / "evidence" / "provider" / "worker-deploy.json"
    evidence = None
    if provider_path.exists():
        try:
            evidence = json.loads(provider_path.read_text(encoding="utf-8"))
        except Exception as exc:
            raise AccessReviewError("ACCESS_REVIEW_PROVIDER_EVIDENCE_INVALID") from exc
    explicit = explicit.rstrip("/") if explicit else None
    if evidence is not None:
        if evidence.get("commit_sha") != head:
            raise AccessReviewError("ACCESS_REVIEW_PROVIDER_HEAD_MISMATCH")
        origin = _provider_origin_from(evidence)
        if not origin:
            raise AccessReviewError("ACCESS_REVIEW_PROVIDER_ORIGIN_MISSING")
        if explicit and explicit != origin:
            raise AccessReviewError("ACCESS_REVIEW_BASE_URL_MISMATCH")
        return origin
    if not explicit or not explicit.startswith("https://"):
        raise AccessReviewError("ACCESS_REVIEW_BASE_URL_REQUIRED")
    return explicit


def evidence_for(record: dict[str, Any], head: str, created_at: str | None = None) -> dict[str, Any]:
    kind = record["kind"]
    review = record["review"]
    entity_key = "source_id" if kind == "source" else "provider_id"
    payload_hash = review_payload_hash(record)
    created_at = created_at or dt.datetime.now(dt.timezone.utc).isoformat().replace("+00:00", "Z")
    return {
        "evidence_id": review["evidence_id"],
        "gate_id": "TERMS_AND_PRIVACY_REVIEW" if kind == "source" else "ACCESS_BASIS_REVIEW",
        "spec_version": "1.3",
        "commit_sha": head,
        "test_report_hash": payload_hash,
        "provider_readback": {"access_review": {
            "kind": kind,
            "entity_id": review[entity_key],
            "review_id": review["review_id"],
            "human_review_payload_sha256": payload_hash,
        }},
        "unresolved_items": [],
        "created_at": created_at,
    }


@dataclass
class HttpResult:
    status: int
    data: Any


class SignedHttpTransport:
    def __init__(self, base_url: str, key_id: str, secret: str, timeout: float = 15.0):
        self.base_url = base_url.rstrip("/")
        self.key_id = key_id
        self.secret = secret
        self.timeout = timeout

    def _open(self, request: urllib.request.Request) -> HttpResult:
        try:
            with urllib.request.urlopen(request, timeout=self.timeout) as response:
                raw = response.read().decode("utf-8")
                return HttpResult(response.status, json.loads(raw) if raw else None)
        except urllib.error.HTTPError as exc:
            raw = exc.read().decode("utf-8", errors="replace")
            try:
                data = json.loads(raw) if raw else None
            except json.JSONDecodeError:
                data = {"error": raw}
            return HttpResult(exc.code, data)
        except (urllib.error.URLError, TimeoutError, socket.timeout, ConnectionError, OSError) as exc:
            raise TransportUnknown("ACCESS_REVIEW_TRANSPORT_UNKNOWN") from exc

    def health(self) -> HttpResult:
        return self._open(urllib.request.Request(self.base_url + "/health", method="GET"))

    def post(self, path: str, payload: dict[str, Any]) -> HttpResult:
        body = _compact(payload)
        timestamp = str(int(dt.datetime.now(dt.timezone.utc).timestamp() * 1000))
        nonce = secrets.token_urlsafe(18).replace("-", "_")[:24]
        body_hash = _sha256_text(body)
        canonical = "\n".join([self.key_id, timestamp, nonce, "POST", path, body_hash])
        signature = hmac.new(self.secret.encode("utf-8"), canonical.encode("utf-8"), hashlib.sha256).hexdigest()
        request = urllib.request.Request(
            self.base_url + path,
            data=body.encode("utf-8"),
            method="POST",
            headers={
                "content-type": "application/json",
                "x-fare-key-id": self.key_id,
                "x-fare-timestamp": timestamp,
                "x-fare-nonce": nonce,
                "x-fare-signature": signature,
            },
        )
        return self._open(request)


def verify_health(transport: Any, head: str) -> dict[str, Any]:
    try:
        result = transport.health()
    except TransportUnknown as exc:
        raise AccessReviewError("ACCESS_REVIEW_HEALTH_UNAVAILABLE") from exc
    data = result.data if isinstance(result.data, dict) else {}
    if result.status != 200 or data.get("spec") != "1.3" or data.get("mode") not in ALLOWED_MODES or str(data.get("commit_sha", "")).lower() != head.lower():
        raise AccessReviewError("ACCESS_REVIEW_RUNTIME_IDENTITY_MISMATCH")
    return data


def _audit_readback_request(record: dict[str, Any]) -> dict[str, Any]:
    review = record["review"]
    kind = record["kind"]
    return {"evidence_id": review["evidence_id"], "kind": kind, "entity_id": review["source_id" if kind == "source" else "provider_id"]}


def _review_readback_request(record: dict[str, Any]) -> dict[str, Any]:
    review = record["review"]
    kind = record["kind"]
    return {"kind": kind, "entity_id": review["source_id" if kind == "source" else "provider_id"], "review_id": review["review_id"]}


def audit_state(record: dict[str, Any], head: str, response: HttpResult) -> str:
    if response.status != 200 or not isinstance(response.data, dict):
        raise AccessReviewError(f"ACCESS_REVIEW_AUDIT_READBACK_FAILED:{response.status}")
    actual = response.data.get("evidence")
    if actual is None:
        return "absent"
    expected = evidence_for(record, head, created_at=actual.get("created_at") if isinstance(actual, dict) else None)
    keys = ("evidence_id", "gate_id", "spec_version", "commit_sha", "test_report_hash", "provider_readback", "unresolved_items")
    if isinstance(actual, dict) and all(actual.get(k) == expected.get(k) for k in keys):
        return "exact"
    return "mismatch"


def review_state(record: dict[str, Any], reviewer_key_id: str, response: HttpResult) -> str:
    if response.status != 200 or not isinstance(response.data, dict):
        raise AccessReviewError(f"ACCESS_REVIEW_READBACK_FAILED:{response.status}")
    actual = response.data.get("review")
    if actual is None:
        return "absent"
    review = record["review"]
    payload_hash = review_payload_hash(record)
    kind = record["kind"]
    if kind == "source":
        checks = {name: review[name] for name in SOURCE_FIELDS[-6:]}
        expected = {
            "review_id": review["review_id"], "source_id": review["source_id"], "target_state": review["target_state"],
            "access_basis": review["access_basis"], "terms_snapshot_at": review["terms_snapshot_at"], "evidence_id": review["evidence_id"],
            "payload_sha256": payload_hash, "reviewer_key_id": reviewer_key_id, "checks_json": checks,
        }
        registry = response.data.get("registry") or {}
        registry_ok = all([
            registry.get("source_id") == review["source_id"], registry.get("lifecycle_state") == review["target_state"],
            registry.get("status") == review["target_state"], registry.get("access_basis") == review["access_basis"],
            registry.get("terms_snapshot_at") == review["terms_snapshot_at"], int(registry.get("kill_switch", 1)) == 0,
            registry.get("kill_switch_state") == "CLEAR",
        ])
    else:
        checks = {name: review[name] for name in PROVIDER_FIELDS[-3:]}
        expected = {
            "review_id": review["review_id"], "provider_id": review["provider_id"], "access_basis": review["access_basis"],
            "terms_snapshot_at": review["terms_snapshot_at"], "rate_policy": review["rate_policy"], "evidence_id": review["evidence_id"],
            "payload_sha256": payload_hash, "reviewer_key_id": reviewer_key_id, "checks_json": checks,
        }
        registry = response.data.get("registry") or {}
        registry_ok = all([
            registry.get("provider_id") == review["provider_id"], registry.get("access_basis") == review["access_basis"],
            registry.get("terms_snapshot_at") == review["terms_snapshot_at"], registry.get("rate_policy") == review["rate_policy"],
        ])
    if isinstance(actual, dict) and all(actual.get(k) == v for k, v in expected.items()) and registry_ok:
        return "exact"
    return "mismatch"


def _read_state(transport: Any, path: str, payload: dict[str, Any], classifier: Callable[[HttpResult], str]) -> str:
    try:
        return classifier(transport.post(path, payload))
    except TransportUnknown as exc:
        raise AccessReviewError("ACCESS_REVIEW_READBACK_TRANSPORT_UNKNOWN") from exc


def _mutate_once_then_reconcile(
    transport: Any,
    mutation_path: str,
    mutation_payload: dict[str, Any],
    read_path: str,
    read_payload: dict[str, Any],
    classifier: Callable[[HttpResult], str],
) -> str:
    try:
        result = transport.post(mutation_path, mutation_payload)
    except TransportUnknown:
        state = _read_state(transport, read_path, read_payload, classifier)
        if state == "exact":
            return "reconciled"
        if state == "absent":
            raise AccessReviewError("ACCESS_REVIEW_TRANSPORT_UNKNOWN")
        raise AccessReviewError("ACCESS_REVIEW_READBACK_MISMATCH")
    if not (200 <= result.status < 300):
        if result.status == 409:
            state = _read_state(transport, read_path, read_payload, classifier)
            if state == "exact":
                return "reconciled"
        raise AccessReviewError(f"ACCESS_REVIEW_MUTATION_REJECTED:{mutation_path}:{result.status}")
    state = _read_state(transport, read_path, read_payload, classifier)
    if state != "exact":
        raise AccessReviewError("ACCESS_REVIEW_POST_READBACK_MISMATCH")
    return "written"


def process_record(record: dict[str, Any], head: str, reviewer_key_id: str, transport: Any, dry_run: bool = False) -> dict[str, str]:
    review = record["review"]
    audit_payload = evidence_for(record, head)
    audit_read = _audit_readback_request(record)
    audit_classifier = lambda result: audit_state(record, head, result)
    audit_pre = _read_state(transport, "/audit/evidence/readback", audit_read, audit_classifier)
    if audit_pre == "mismatch":
        raise AccessReviewError("ACCESS_REVIEW_AUDIT_PRESTATE_MISMATCH")
    if audit_pre == "absent":
        if dry_run:
            return {"review_id": review["review_id"], "audit": "would_write", "review": "would_write"}
        audit_action = _mutate_once_then_reconcile(transport, "/audit/evidence", audit_payload, "/audit/evidence/readback", audit_read, audit_classifier)
    else:
        audit_action = "existing"

    review_read = _review_readback_request(record)
    review_classifier = lambda result: review_state(record, reviewer_key_id, result)
    review_pre = _read_state(transport, "/access/reviews/readback", review_read, review_classifier)
    if review_pre == "mismatch":
        raise AccessReviewError("ACCESS_REVIEW_PRESTATE_MISMATCH")
    if review_pre == "absent":
        if dry_run:
            review_action = "would_write"
        else:
            endpoint = "/sources/onboarding/review" if record["kind"] == "source" else "/providers/access/review"
            review_action = _mutate_once_then_reconcile(transport, endpoint, review, "/access/reviews/readback", review_read, review_classifier)
    else:
        review_action = "existing"
    return {"review_id": review["review_id"], "audit": audit_action, "review": review_action}


def run(records: list[dict[str, Any]], head: str, reviewer_key_id: str, transport: Any, dry_run: bool = False) -> list[dict[str, str]]:
    verify_health(transport, head)
    if dry_run:
        # Dry-run remains provider-read-only except for signed nonce bookkeeping on exact readbacks.
        pass
    return [process_record(record, head, reviewer_key_id, transport, dry_run=dry_run) for record in records]


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Submit immutable human-authored source/provider access reviews")
    parser.add_argument("--input", required=True)
    parser.add_argument("--dry-run", action="store_true")
    args = parser.parse_args(argv)
    try:
        _require_clean()
        head = _git_head()
        key_id = os.environ.get("FARE_ACCESS_REVIEWER_KEY_ID", "")
        secret = os.environ.get("FARE_ACCESS_REVIEWER_SECRET", "")
        if not key_id:
            raise AccessReviewError("FARE_ACCESS_REVIEWER_KEY_ID_REQUIRED")
        if len(secret) < 16:
            raise AccessReviewError("FARE_ACCESS_REVIEWER_SECRET_REQUIRED")
        records = load_jsonl(pathlib.Path(args.input))
        base_url = resolve_base_url(head, os.environ.get("FARE_RADAR_BASE_URL"))
        transport = SignedHttpTransport(base_url, key_id, secret)
        result = run(records, head, key_id, transport, dry_run=args.dry_run)
        print(json.dumps({"ok": True, "dry_run": args.dry_run, "commit_sha": head, "base_url": base_url, "results": result}, ensure_ascii=False, indent=2))
        return 0
    except (AccessReviewError, OSError) as exc:
        print(json.dumps({"ok": False, "error": str(exc)}, ensure_ascii=False), file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: src/access_review.ts
import type { D1Database } from "./types.js";

const PROVIDER_ACCESS_BASES=new Set(["OFFICIAL_API","PARTNER_CONTRACT","AIRLINE_DIRECT","MANUAL_ORACLE","PUBLIC_PAGE_MONITOR"]);
const SHA256=/^[a-f0-9]{64}$/i;
const COMMIT=/^[a-f0-9]{40}$/i;

async function sha256Hex(text:string){const b=new TextEncoder().encode(text);const h=await crypto.subtle.digest("SHA-256",b);return [...new Uint8Array(h)].map(x=>x.toString(16).padStart(2,"0")).join("");}
function iso(v:unknown){return typeof v==="string"&&v!=="RECHECK_REQUIRED"&&Number.isFinite(Date.parse(v));}
function explicitBool(v:unknown){return typeof v==="boolean";}
function parseJson(v:unknown){if(typeof v!=="string")return v;try{return JSON.parse(v);}catch{return null;}}
function unresolvedEmpty(v:unknown){const x=parseJson(v);return Array.isArray(x)&&x.length===0;}

export interface SourceAccessReview {
  review_id:string; source_id:string; target_state:"SHADOW"|"ENABLED"; access_basis:string;
  terms_snapshot_at:string; evidence_id:string; access_basis_valid:boolean; privacy_review_pass:boolean;
  parser_contract_pass:boolean; provenance_hash_pass:boolean; rate_budget_pass:boolean; shadow_pass:boolean;
}
export interface ProviderAccessReview {
  review_id:string; provider_id:string; access_basis:string; terms_snapshot_at:string; rate_policy:string;
  evidence_id:string; access_basis_valid:boolean; terms_review_pass:boolean; rate_policy_review_pass:boolean;
}
export type AccessReviewKind="source"|"provider";

function sourceCanonical(r:SourceAccessReview){return JSON.stringify({review_id:r.review_id,source_id:r.source_id,target_state:r.target_state,access_basis:r.access_basis,terms_snapshot_at:r.terms_snapshot_at,evidence_id:r.evidence_id,access_basis_valid:r.access_basis_valid,privacy_review_pass:r.privacy_review_pass,parser_contract_pass:r.parser_contract_pass,provenance_hash_pass:r.provenance_hash_pass,rate_budget_pass:r.rate_budget_pass,shadow_pass:r.shadow_pass});}
function providerCanonical(r:ProviderAccessReview){return JSON.stringify({review_id:r.review_id,provider_id:r.provider_id,access_basis:r.access_basis,terms_snapshot_at:r.terms_snapshot_at,rate_policy:r.rate_policy,evidence_id:r.evidence_id,access_basis_valid:r.access_basis_valid,terms_review_pass:r.terms_review_pass,rate_policy_review_pass:r.rate_policy_review_pass});}
export async function sourceAccessReviewPayloadHash(r:SourceAccessReview){return sha256Hex(sourceCanonical(r));}
export async function providerAccessReviewPayloadHash(r:ProviderAccessReview){return sha256Hex(providerCanonical(r));}

export function runtimeCommitOrNull(v:string|undefined){return v&&COMMIT.test(v)?v.toLowerCase():null;}
export function validateAccessAuditEvidenceInput(e:any,runtimeCommitSha:string){
  if(!COMMIT.test(runtimeCommitSha))throw new Error("RUNTIME_COMMIT_UNAVAILABLE");
  if(!e||!e.evidence_id||e.spec_version!=="1.3"||String(e.commit_sha??"").toLowerCase()!==runtimeCommitSha.toLowerCase()||!SHA256.test(String(e.test_report_hash??""))||!iso(e.created_at)||!unresolvedEmpty(e.unresolved_items))throw new Error("ACCESS_AUDIT_EVIDENCE_INVALID");
  const binding=e.provider_readback?.access_review;
  if(!binding||!['source','provider'].includes(binding.kind)||!binding.entity_id||!binding.review_id||binding.human_review_payload_sha256!==e.test_report_hash)throw new Error("ACCESS_AUDIT_BINDING_INVALID");
  const expectedGate=binding.kind==="source"?"TERMS_AND_PRIVACY_REVIEW":"ACCESS_BASIS_REVIEW";
  if(e.gate_id!==expectedGate)throw new Error("ACCESS_AUDIT_STAGE_INVALID");
  return {kind:binding.kind as AccessReviewKind,entity_id:String(binding.entity_id),review_id:String(binding.review_id),payload_sha256:String(binding.human_review_payload_sha256)};
}

async function requireBoundEvidence(db:D1Database,input:{evidence_id:string;kind:AccessReviewKind;entity_id:string;review_id:string;payload_sha256:string;runtime_commit_sha:string}){
  const ev=await db.prepare("SELECT gate_id,spec_version,commit_sha,test_report_hash,provider_readback,unresolved_items FROM audit_evidence WHERE evidence_id=?").bind(input.evidence_id).first<any>();
  const expectedGate=input.kind==="source"?"TERMS_AND_PRIVACY_REVIEW":"ACCESS_BASIS_REVIEW";
  if(!ev||ev.gate_id!==expectedGate||ev.spec_version!=="1.3"||String(ev.commit_sha??"").toLowerCase()!==input.runtime_commit_sha.toLowerCase()||ev.test_report_hash!==input.payload_sha256||!unresolvedEmpty(ev.unresolved_items))throw new Error("ACCESS_REVIEW_EVIDENCE_INVALID");
  const b=parseJson(ev.provider_readback)?.access_review;
  if(!b||b.kind!==input.kind||b.entity_id!==input.entity_id||b.review_id!==input.review_id||b.human_review_payload_sha256!==input.payload_sha256)throw new Error("ACCESS_REVIEW_EVIDENCE_BINDING_MISMATCH");
}

export async function applyAccessSourceReview(db:D1Database,r:SourceAccessReview,ctx:{nowIso:string;reviewerKeyId:string;runtimeCommitSha:string}){
  if(!r||!r.review_id||!r.source_id||!r.evidence_id||!r.access_basis?.trim()||!iso(r.terms_snapshot_at)||!['SHADOW','ENABLED'].includes(r.target_state)||![r.access_basis_valid,r.privacy_review_pass,r.parser_contract_pass,r.provenance_hash_pass,r.rate_budget_pass,r.shadow_pass].every(explicitBool))throw new Error("SOURCE_REVIEW_INVALID");
  if(!ctx.reviewerKeyId||!COMMIT.test(ctx.runtimeCommitSha))throw new Error("RUNTIME_COMMIT_UNAVAILABLE");
  const payloadHash=await sourceAccessReviewPayloadHash(r);
  const prior=await db.prepare("SELECT payload_sha256,reviewer_key_id FROM source_onboarding_reviews WHERE review_id=?").bind(r.review_id).first<any>();
  if(prior){if(prior.payload_sha256!==payloadHash||prior.reviewer_key_id!==ctx.reviewerKeyId)throw new Error("SOURCE_REVIEW_IDEMPOTENCY_CONFLICT");return {source_id:r.source_id,state:r.target_state,payload_sha256:payloadHash,reviewer_key_id:prior.reviewer_key_id,idempotent:true};}
  const src=await db.prepare("SELECT lifecycle_state,parser FROM source_registry WHERE source_id=?").bind(r.source_id).first<any>();
  if(!src)throw new Error("SOURCE_NOT_FOUND");
  await requireBoundEvidence(db,{evidence_id:r.evidence_id,kind:"source",entity_id:r.source_id,review_id:r.review_id,payload_sha256:payloadHash,runtime_commit_sha:ctx.runtimeCommitSha});
  if(!r.access_basis_valid||!r.privacy_review_pass||!r.parser_contract_pass)throw new Error("SOURCE_SHADOW_GATES_INCOMPLETE");
  if(!src.parser)throw new Error("SOURCE_CONTRACT_INCOMPLETE");
  if(r.target_state==="ENABLED"){
    if(src.lifecycle_state!=="SHADOW")throw new Error("SOURCE_MUST_PASS_SHADOW_FIRST");
    if(!r.provenance_hash_pass||!r.rate_budget_pass||!r.shadow_pass)throw new Error("SOURCE_ENABLE_GATES_INCOMPLETE");
  }
  const checks={access_basis_valid:r.access_basis_valid,privacy_review_pass:r.privacy_review_pass,parser_contract_pass:r.parser_contract_pass,provenance_hash_pass:r.provenance_hash_pass,rate_budget_pass:r.rate_budget_pass,shadow_pass:r.shadow_pass};
  await db.batch([
    db.prepare("INSERT INTO source_onboarding_reviews(review_id,source_id,target_state,terms_snapshot_at,evidence_id,checks_json,payload_sha256,created_at,access_basis,reviewer_key_id) VALUES(?,?,?,?,?,?,?,?,?,?)").bind(r.review_id,r.source_id,r.target_state,r.terms_snapshot_at,r.evidence_id,JSON.stringify(checks),payloadHash,ctx.nowIso,r.access_basis,ctx.reviewerKeyId),
    db.prepare("UPDATE source_registry SET lifecycle_state=?,status=?,access_basis=?,terms_snapshot_at=?,kill_switch=0,kill_switch_state='CLEAR' WHERE source_id=?").bind(r.target_state,r.target_state,r.access_basis,r.terms_snapshot_at,r.source_id)
  ]);
  return {source_id:r.source_id,state:r.target_state,payload_sha256:payloadHash,reviewer_key_id:ctx.reviewerKeyId,idempotent:false};
}

export async function applyProviderAccessReview(db:D1Database,r:ProviderAccessReview,ctx:{nowIso:string;reviewerKeyId:string;runtimeCommitSha:string}){
  if(!r||!r.review_id||!r.provider_id||!r.evidence_id||!PROVIDER_ACCESS_BASES.has(r.access_basis)||!iso(r.terms_snapshot_at)||!r.rate_policy?.trim()||![r.access_basis_valid,r.terms_review_pass,r.rate_policy_review_pass].every(explicitBool))throw new Error("PROVIDER_ACCESS_REVIEW_INVALID");
  if(!ctx.reviewerKeyId||!COMMIT.test(ctx.runtimeCommitSha))throw new Error("RUNTIME_COMMIT_UNAVAILABLE");
  if(!r.access_basis_valid||!r.terms_review_pass||!r.rate_policy_review_pass)throw new Error("PROVIDER_ACCESS_REVIEW_INCOMPLETE");
  const payloadHash=await providerAccessReviewPayloadHash(r);
  const prior=await db.prepare("SELECT payload_sha256,reviewer_key_id FROM provider_access_reviews WHERE review_id=?").bind(r.review_id).first<any>();
  if(prior){if(prior.payload_sha256!==payloadHash||prior.reviewer_key_id!==ctx.reviewerKeyId)throw new Error("PROVIDER_ACCESS_REVIEW_IDEMPOTENCY_CONFLICT");return {provider_id:r.provider_id,payload_sha256:payloadHash,reviewer_key_id:prior.reviewer_key_id,idempotent:true};}
  const provider=await db.prepare("SELECT provider_id FROM provider_access_registry WHERE provider_id=?").bind(r.provider_id).first<any>();
  if(!provider)throw new Error("PROVIDER_NOT_FOUND");
  await requireBoundEvidence(db,{evidence_id:r.evidence_id,kind:"provider",entity_id:r.provider_id,review_id:r.review_id,payload_sha256:payloadHash,runtime_commit_sha:ctx.runtimeCommitSha});
  const checks={access_basis_valid:r.access_basis_valid,terms_review_pass:r.terms_review_pass,rate_policy_review_pass:r.rate_policy_review_pass};
  await db.batch([
    db.prepare("INSERT INTO provider_access_reviews(review_id,provider_id,access_basis,terms_snapshot_at,rate_policy,evidence_id,checks_json,payload_sha256,reviewer_key_id,created_at) VALUES(?,?,?,?,?,?,?,?,?,?)").bind(r.review_id,r.provider_id,r.access_basis,r.terms_snapshot_at,r.rate_policy,r.evidence_id,JSON.stringify(checks),payloadHash,ctx.reviewerKeyId,ctx.nowIso),
    db.prepare("UPDATE provider_access_registry SET access_basis=?,terms_snapshot_at=?,rate_policy=? WHERE provider_id=?").bind(r.access_basis,r.terms_snapshot_at,r.rate_policy,r.provider_id)
  ]);
  return {provider_id:r.provider_id,payload_sha256:payloadHash,reviewer_key_id:ctx.reviewerKeyId,idempotent:false};
}

export async function readAccessReview(db:D1Database,input:{kind:AccessReviewKind;entity_id:string;review_id?:string|null}){
  if(!input||!['source','provider'].includes(input.kind)||!input.entity_id)throw new Error("ACCESS_REVIEW_READBACK_INVALID");
  if(input.kind==="source"){
    const review=input.review_id?await db.prepare("SELECT review_id,source_id,target_state,access_basis,terms_snapshot_at,evidence_id,checks_json,payload_sha256,reviewer_key_id,created_at FROM source_onboarding_reviews WHERE source_id=? AND review_id=?").bind(input.entity_id,input.review_id).first<any>():await db.prepare("SELECT review_id,source_id,target_state,access_basis,terms_snapshot_at,evidence_id,checks_json,payload_sha256,reviewer_key_id,created_at FROM source_onboarding_reviews WHERE source_id=? ORDER BY created_at DESC,review_id DESC LIMIT 1").bind(input.entity_id).first<any>();
    const registry=await db.prepare("SELECT source_id,lifecycle_state,status,access_basis,terms_snapshot_at,kill_switch,kill_switch_state FROM source_registry WHERE source_id=?").bind(input.entity_id).first<any>();
    return {kind:"source",review:review?{...review,checks_json:parseJson(review.checks_json)}:null,registry};
  }
  const review=input.review_id?await db.prepare("SELECT review_id,provider_id,access_basis,terms_snapshot_at,rate_policy,evidence_id,checks_json,payload_sha256,reviewer_key_id,created_at FROM provider_access_reviews WHERE provider_id=? AND review_id=?").bind(input.entity_id,input.review_id).first<any>():await db.prepare("SELECT review_id,provider_id,access_basis,terms_snapshot_at,rate_policy,evidence_id,checks_json,payload_sha256,reviewer_key_id,created_at FROM provider_access_reviews WHERE provider_id=? ORDER BY created_at DESC,review_id DESC LIMIT 1").bind(input.entity_id).first<any>();
  const registry=await db.prepare("SELECT provider_id,access_basis,terms_snapshot_at,rate_policy,kill_switch_state,connector_state FROM provider_access_registry WHERE provider_id=?").bind(input.entity_id).first<any>();
  return {kind:"provider",review:review?{...review,checks_json:parseJson(review.checks_json)}:null,registry};
}
